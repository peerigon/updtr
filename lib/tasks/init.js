"use strict";

const createSequence = require("../createSequence");
const createUpdateTask = require("../createUpdateTask");
const filterUpdateTask = require("../filterUpdateTask");

function init(instance) {
    const baseEvent = instance.config;
    let sequence;

    return Promise.resolve()
        .then(() => {
            sequence = createSequence(instance, baseEvent);

            return sequence.exec(
                "installMissing",
                instance.cmds.installMissing()
            );
        })
        .then(() =>
            sequence.exec("collect", instance.cmds.outdated()).catch(err => {
                if (err.code > 1) {
                    throw err;
                }
            }))
        .then(() => {
            const stdout = sequence.stdouts.outdated.trim();

            if (stdout.length === 0) {
                // The result of the init task is an array of update tasks.
                // Hence, we're returning an empty array when there's nothing outdated.
                return [];
            }

            return instance.parse
                .outdated(stdout)
                .map(outdated => createUpdateTask(instance, outdated))
                .filter(updateTask =>
                    filterUpdateTask(updateTask, instance.config));
        })
        .then(updateTasks => {
            instance.emit("collectDone", {
                updateTasks,
            });

            return updateTasks;
        });
}

module.exports = init;
