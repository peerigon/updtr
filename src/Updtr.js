import EventEmitter from "events";
import path from "path";
import fs from "./util/fs";
import {
    USE_OPTIONS,
    UPDATE_TO_OPTIONS,
    SAVE_OPTIONS,
} from "./constants/config";
import exec from "./exec/exec";
import cmds from "./exec/cmds";
import parse from "./exec/parse";
import {
    RequiredOptionMissingError,
    OptionValueNotSupportedError,
    YarnWithCustomRegistryError,
} from "./errors";

// node v4 has no dedicated constants object.
// Remove this if node v4 is not supported anymore.
const FS_CONSTANTS = fs.constants === undefined ? fs : fs.constants;

function checkCwd(cwd) {
    if (typeof cwd !== "string") {
        throw new RequiredOptionMissingError("cwd", cwd);
    }
}

function checkUse(use) {
    if (USE_OPTIONS.indexOf(use) === -1) {
        throw new OptionValueNotSupportedError("use", use);
    }
}

function checkUpdateTo(updateTo) {
    if (UPDATE_TO_OPTIONS.indexOf(updateTo) === -1) {
        throw new OptionValueNotSupportedError("updateTo", updateTo);
    }
}

function checkSave(save) {
    if (SAVE_OPTIONS.indexOf(save) === -1) {
        throw new OptionValueNotSupportedError("save", save);
    }
}

function checkForYarnWithCustomReg(packageManager, registry) {
    if (packageManager === "yarn" && registry !== undefined) {
        throw new YarnWithCustomRegistryError();
    }
}

export default class Updtr extends EventEmitter {
    /**
     * The config passed-in here should look identically to the CLI config.
     * Dash-cased properties should be renamed to camelCased.
     * The goal is to replicate the API of the CLI as close as possible so users don't
     * have to guess the options.
     *
     * @param {UpdtrConfig} config
     */
    constructor(config) {
        super();

        const cwd = config.cwd;
        const registry = config.registry;
        const packageManager = config.use === undefined ?
            USE_OPTIONS[0] :
            config.use;
        const updateTo = config.updateTo === undefined ?
            UPDATE_TO_OPTIONS[0] :
            config.updateTo;
        const exclude = Array.isArray(config.exclude) === true ?
            config.exclude :
            [];
        const save = config.save === undefined ? SAVE_OPTIONS[0] : config.save;

        checkCwd(cwd);
        checkUse(packageManager);
        checkForYarnWithCustomReg(packageManager, registry);
        checkUpdateTo(updateTo);
        checkSave(save);

        this.config = {
            cwd,
            use: packageManager,
            exclude,
            test: config.test,
            registry,
            updateTo,
            save,
        };
        this.cmds = cmds[packageManager];
        this.parse = parse[packageManager];

        if (typeof config.test === "string") {
            this.cmds = {
                ...this.cmds,
                test: () => config.test,
            };
        }
    }
    async canAccessPackageJson() {
        let result = true;

        try {
            await fs.access(
                path.join(this.config.cwd, "package.json"),
                FS_CONSTANTS.R_OK | FS_CONSTANTS.W_OK // eslint-disable-line no-bitwise
            );
        } catch (err) {
            result = false;
        }

        return result;
    }
    exec(cmd) {
        return exec(this.config.cwd, cmd);
    }
    readFile(filenameInCwd) {
        return fs.readFile(path.join(this.config.cwd, filenameInCwd), "utf8");
    }
    writeFile(filenameInCwd, contents) {
        return fs.writeFile(
            path.join(this.config.cwd, filenameInCwd),
            contents
        );
    }
    dispose() {
        this.removeAllListeners();
    }
}
