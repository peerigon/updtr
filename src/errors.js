// Errors are not extendable in node v4.
// Remove this if node v4 is not supported anymore.
import Error from "es6-error";

export class PackageJsonNoAccessError extends Error {
    constructor(dir) {
        super(`Cannot access package.json in ${ dir }`);
    }
}

export class RequiredOptionMissingError extends Error {
    constructor(optionName, optionValue) {
        super(
            `Required option ${ optionName } is missing. Instead received ${ optionValue }`
        );
    }
}

export class OptionValueNotSupportedError extends Error {
    constructor(optionName, unsupportedValue) {
        super(`Unsupported value ${ unsupportedValue } for option ${ optionName }`);
    }
}

export class YarnWithCustomRegistryError extends Error {
    constructor() {
        super(
            "yarn does not support custom registries yet. Please use a .npmrc file to achieve this"
        );
    }
}
