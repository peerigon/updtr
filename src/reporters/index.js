import dense from "./dense";
import none from "./none";

// The first property here is the default reporter
export default { dense, none };
