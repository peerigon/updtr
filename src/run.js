import init from "./tasks/init";
import sequentialUpdate from "./tasks/sequentialUpdate";
import splitUpdateTasks from "./tasks/util/splitUpdateTasks";
import batchUpdate from "./tasks/batchUpdate";
import createUpdateResult from "./tasks/util/createUpdateResult";
import finish from "./tasks/finish";
import updatePackageJson from "./tasks/updatePackageJson";

async function runUpdateTasks(updtr, updateTasks) {
    const results = [];
    const { breaking, nonBreaking } = splitUpdateTasks(updateTasks);
    const sequentialUpdateTasks = breaking.slice();
    let batchSuccess; // can be undefined, true or false
    let batchUpdateFailure;

    // Run batch update if we have more than one non-breaking update
    // If the batch update fails, it will roll back all modules except the first one.
    // This way we can skip one install command since we will run the sequential update for it anyway.
    if (nonBreaking.length > 1) {
        batchSuccess = await batchUpdate(updtr, nonBreaking);
    }

    if (batchSuccess === true) {
        results.push(
            ...nonBreaking.map(updateTask =>
                createUpdateResult(updateTask, true)
            )
        );
    } else {
        sequentialUpdateTasks.unshift(...nonBreaking);
        // If batchSuccess is false, we have actually executed the batch update and it returned false
        if (batchSuccess === false) {
            batchUpdateFailure = createUpdateResult(nonBreaking[0], false);
        }
    }

    // Run sequential update for all breaking updates and non-breaking batch updates that failed
    results.push(
        ...(await sequentialUpdate(
            updtr,
            sequentialUpdateTasks,
            batchUpdateFailure
        ))
    );

    return finish(updtr, results);
}

export default (async function run(updtr) {
    const results = [];

    updtr.emit("start", {
        config: updtr.config,
    });

    const { updateTasks } = await init(updtr);

    if (updateTasks.length > 0) {
        results.push(...(await runUpdateTasks(updtr, updateTasks)));
        await updatePackageJson(updtr, results);
    }

    updtr.emit("end", {
        config: updtr.config,
        results,
    });

    return results;
});
