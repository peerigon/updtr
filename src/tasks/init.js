import Sequence from "./util/Sequence";
import createUpdateTask from "./util/createUpdateTask";
import filterUpdateTask from "./util/filterUpdateTask";
import { PackageJsonNoAccessError } from "../errors";

function getUpdateTasksFromStdout(updtr, outdatedCmd, stdout) {
    if (stdout.length === 0) {
        // When there is not stdout, there is nothing to update
        return [];
    }

    return updtr.parse
        .outdated(stdout, outdatedCmd)
        .map(outdated => createUpdateTask(outdated, updtr.config));
}

export default (async function init(updtr) {
    const baseEvent = { config: updtr.config };
    const outdatedCmd = updtr.cmds.outdated();
    const sequence = new Sequence("init", updtr, baseEvent);
    let stdout;

    sequence.start();

    if ((await updtr.canAccessPackageJson()) === false) {
        throw new PackageJsonNoAccessError(updtr.config.cwd);
    }

    await sequence.exec(
        "install-missing",
        updtr.cmds.installMissing({
            registry: updtr.config.registry,
        })
    );

    try {
        stdout = (await sequence.exec("collect", outdatedCmd)).stdout;
    } catch (err) {
        // npm exits with zero code 1 when there are outdated dependencies
        // We don't check for the package manager here because yarn might change their
        // behavior in the future to be npm-compatible.
        if (err.code > 1) {
            throw err;
        }

        stdout = err.stdout;
    }

    const allUpdateTasks = getUpdateTasksFromStdout(
        updtr,
        outdatedCmd,
        stdout.trim()
    );
    const filterResults = allUpdateTasks.map(updateTask =>
        filterUpdateTask(updateTask, updtr.config));
    const result = {
        updateTasks: allUpdateTasks.filter(
            (updateTask, index) => filterResults[index] === null
        ),
        excluded: allUpdateTasks.reduce(
            (excluded, updateTask, index) => {
                const reason = filterResults[index];

                if (reason === null) {
                    return excluded;
                }

                return excluded.concat({
                    ...updateTask,
                    reason,
                });
            },
            []
        ),
    };

    sequence.end(result);

    return result;
});
