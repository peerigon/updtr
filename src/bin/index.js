import argv from "./argv";
import { create, run } from "./..";
import reporters from "../reporters";

async function start() {
    const cwd = process.cwd();
    const config = { ...argv };
    const reporterConfig = {
        stream: process.stdout,
        testStdout: argv.testStdout,
    };
    const reporter = reporters[argv.reporter];

    config.cwd = cwd;

    const updtr = create(config);

    reporter(updtr, reporterConfig);
    try {
        await run(updtr);
    } catch (err) {
        updtr.emit("error", err);
    }
}

start();
