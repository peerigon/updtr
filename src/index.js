import run from "./run";
import Updtr from "./Updtr";
import * as errors from "./errors";

export function create(config) {
    return new Updtr(config);
}

export { errors, run };
