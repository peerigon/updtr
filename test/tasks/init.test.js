import init from "../../src/tasks/init";
import FakeUpdtr from "../helpers/FakeUpdtr";
import {
    execError,
    npmNoOutdated,
    npmOutdated,
    yarnOutdated,
    errorExecInstallMissing,
    errorExecOutdated,
    errorParseOutdated,
} from "../fixtures/execResults";
import { PackageJsonNoAccessError } from "../../src/errors";

describe("init()", () => {
    describe("when there are no outdated dependencies", () => {
        it("should emit expected events and execute expected commands", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmNoOutdated;

            await init(updtr);

            expect(updtr.exec.args).toMatchSnapshot();
            expect(updtr.emit.args).toMatchSnapshot();
        });
        it("should return the results as emitted with the 'init/end' event", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmNoOutdated;

            const returnedResults = await init(updtr);
            const [name, endEvent] = updtr.emit.args.pop();

            expect(name).toBe("init/end");
            // The emitted event has additional properties like the config
            expect(endEvent).toEqual(expect.objectContaining(returnedResults));
        });
    });
    describe("when there are outdated dependencies", () => {
        describe("using npm", () => {
            it("should emit expected events and execute expected commands", async () => {
                const updtr = new FakeUpdtr();

                updtr.execResults = npmOutdated;

                await init(updtr);

                expect(updtr.exec.args).toMatchSnapshot();
                expect(updtr.emit.args).toMatchSnapshot();
            });
            it("should return the results as emitted with the 'init/end' event", async () => {
                const updtr = new FakeUpdtr();

                updtr.execResults = npmOutdated;

                const returnedResults = await init(updtr);
                const [name, endEvent] = updtr.emit.args.pop();

                expect(name).toBe("init/end");
                // The emitted event has additional properties like the config
                expect(endEvent).toEqual(
                    expect.objectContaining(returnedResults)
                );
            });
        });
        describe("using yarn", () => {
            it("should emit expected events and execute expected commands", async () => {
                const updtr = new FakeUpdtr({
                    use: "yarn",
                });

                updtr.execResults = yarnOutdated;

                await init(updtr);

                expect(updtr.exec.args).toMatchSnapshot();
                expect(updtr.emit.args).toMatchSnapshot();
            });
            // We don't test for everything here because we assume that the rest works the same as with npm
        });
    });
    describe("when there are excluded dependencies", () => {
        it("should emit expected events and execute expected commands", async () => {
            const updtr = new FakeUpdtr({
                exclude: ["updtr-test-module-1", "updtr-test-module-2"],
            });

            updtr.execResults = npmOutdated;

            await init(updtr);

            expect(updtr.exec.args).toMatchSnapshot();
            expect(updtr.emit.args).toMatchSnapshot();
        });
    });
    describe("when there is no package.json in the cwd", () => {
        it("should throw a PackageJsonNoAccessError", async () => {
            const updtr = new FakeUpdtr();
            let givenErr;

            updtr.canAccessPackageJson.resolves(false);

            try {
                await init(updtr);
            } catch (err) {
                givenErr = err;
            }
            expect(givenErr).toBeInstanceOf(PackageJsonNoAccessError);
        });
    });
    describe("unexpected errors", () => {
        it("should fail when installMissing cmd exits with a non-zero exit code", async () => {
            const updtr = new FakeUpdtr();
            let givenErr;

            updtr.execResults = errorExecInstallMissing;

            try {
                await init(updtr);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBe(execError);
        });
        it("should fail when the outdated cmd exits with an exit code above 1", async () => {
            const updtr = new FakeUpdtr();
            let givenErr;

            updtr.execResults = errorExecOutdated;

            try {
                await init(updtr);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBe(execError);
        });
        it("should fail with a parse error when the stdout could not be parsed", async () => {
            const updtr = new FakeUpdtr();
            let givenErr;

            updtr.execResults = errorParseOutdated;

            try {
                await init(updtr);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBeInstanceOf(SyntaxError);
            expect(givenErr.message).toMatch(
                /Error when trying to parse stdout from command 'npm outdated --json --depth=0': Unexpected token N/
            );
        });
    });
});
