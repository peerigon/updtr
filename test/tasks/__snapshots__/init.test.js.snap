// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init() when there are excluded dependencies should emit expected events and execute expected commands 1`] = `
Array [
  Array [
    "npm install",
  ],
  Array [
    "npm outdated --json --depth=0",
  ],
]
`;

exports[`init() when there are excluded dependencies should emit expected events and execute expected commands 2`] = `
Array [
  Array [
    "init/start",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [
          "updtr-test-module-1",
          "updtr-test-module-2",
        ],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/install-missing",
    Object {
      "cmd": "npm install",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [
          "updtr-test-module-1",
          "updtr-test-module-2",
        ],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/collect",
    Object {
      "cmd": "npm outdated --json --depth=0",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [
          "updtr-test-module-1",
          "updtr-test-module-2",
        ],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/end",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [
          "updtr-test-module-1",
          "updtr-test-module-2",
        ],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
      "excluded": Array [
        Object {
          "name": "updtr-test-module-1",
          "reason": "excluded",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "reason": "excluded",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
      "updateTasks": Array [],
    },
  ],
]
`;

exports[`init() when there are no outdated dependencies should emit expected events and execute expected commands 1`] = `
Array [
  Array [
    "npm install",
  ],
  Array [
    "npm outdated --json --depth=0",
  ],
]
`;

exports[`init() when there are no outdated dependencies should emit expected events and execute expected commands 2`] = `
Array [
  Array [
    "init/start",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/install-missing",
    Object {
      "cmd": "npm install",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/collect",
    Object {
      "cmd": "npm outdated --json --depth=0",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/end",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
      "excluded": Array [],
      "updateTasks": Array [],
    },
  ],
]
`;

exports[`init() when there are outdated dependencies using npm should emit expected events and execute expected commands 1`] = `
Array [
  Array [
    "npm install",
  ],
  Array [
    "npm outdated --json --depth=0",
  ],
]
`;

exports[`init() when there are outdated dependencies using npm should emit expected events and execute expected commands 2`] = `
Array [
  Array [
    "init/start",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/install-missing",
    Object {
      "cmd": "npm install",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/collect",
    Object {
      "cmd": "npm outdated --json --depth=0",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
    },
  ],
  Array [
    "init/end",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "npm",
      },
      "excluded": Array [],
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
]
`;

exports[`init() when there are outdated dependencies using yarn should emit expected events and execute expected commands 1`] = `
Array [
  Array [
    "yarn",
  ],
  Array [
    "yarn outdated --json --flat",
  ],
]
`;

exports[`init() when there are outdated dependencies using yarn should emit expected events and execute expected commands 2`] = `
Array [
  Array [
    "init/start",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "yarn",
      },
    },
  ],
  Array [
    "init/install-missing",
    Object {
      "cmd": "yarn",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "yarn",
      },
    },
  ],
  Array [
    "init/collect",
    Object {
      "cmd": "yarn outdated --json --flat",
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "yarn",
      },
    },
  ],
  Array [
    "init/end",
    Object {
      "config": Object {
        "cwd": "/updtr/test/cwd",
        "exclude": Array [],
        "registry": undefined,
        "save": "smart",
        "test": undefined,
        "updateTo": "latest",
        "use": "yarn",
      },
      "excluded": Array [],
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
]
`;
